### ISSUE TITLE
Node::scroll_area should check if the element is potentially scrollable

### ISSUE THREAD

**Issue by danlrobertson** [servo/servo#10947]
[`Node::scroll_area`](https://github.com/servo/servo/blob/master/components/script/dom/node.rs#L585-L615) should check if the element is [`potentially_scrollable`](https://github.com/servo/servo/blob/master/components/script/dom/element.rs#L179-184) for the second condition of [line 607](https://github.com/servo/servo/blob/master/components/script/dom/node.rs#L607)

relevant specs:
[scrollWidth](https://drafts.csswg.org/cssom-view/#dom-element-scrollwidth): Step 5
[scrollHeight](https://drafts.csswg.org/cssom-view/#dom-element-scrollheight): Step 5
[scrollLeft](https://drafts.csswg.org/cssom-view/#dom-element-scrollleft): Step 7
[scrollTop](https://drafts.csswg.org/cssom-view/#dom-element-scrolltop): Step 7
[potentially scrollable](https://drafts.csswg.org/cssom-view/#potentially-scrollable)

**Comment by broesamle** [comment]
I'd like to work on this one.

**Comment by danlrobertson** [comment]
Awesome! Feel free to ping me if you have any questions!

**Comment by broesamle** [comment]
To me `/web-platform-tests/cssom-view/scrollingElement.html` is a relevant test here or, at least is a good starting point. 

For me both subtests I see a fail as expected but with a strange message in the window: `URL.createObjectURL is not a function. 

Any further suggestions regarding tests?

**Comment by jdm** [comment]
We don't implement the `createObjectURL` API yet; that's part of one of the GSoC projects that were accepted for Servo this year.

**Comment by danlrobertson** [comment]
As of this moment `Element::ScrollHeight` and `Element::ScrollWidth` are the only places `Node::scroll_area` is used. Iif I'm not mistaken neither of these functions are used in `test-wpt`. The most relevant tests I'm aware of are `/css-tests/cssom-view-1_dev/scrollWidthHeight*`. You can run them with something like the following.

```
./mach test-css tests/wpt/css-tests/cssom-view-1_dev/html/scrollWidthHeight*
```

The outcome of these tests shouldn't change since none of the elements have `overflow-(x|y): visible`. Perhapse a test should be added for an element with `overflow-(x|y): visible`?

**Comment by broesamle** [comment]
yes, that was the other candidate... but I need quirk mode and a body element.

It will need the visibility (→pot. srollable), and a different result from the window vs the scrolling area.

Ill try to write sth that would be sensitive to the change...


Martin Br:osamle     m-broe.de

-------- Ursprüngliche Nachricht --------
Von: Dan Robertson <notifications@github.com> 
Datum: 04.05.2016  22:48  (GMT+01:00) 
An: servo/servo <servo@noreply.github.com> 
Cc: "Martin Br:osamle" <info@m-broe.de>,Comment <comment@noreply.github.com> 
Betreff: Re: [servo/servo] Node::scroll_area should check if the element is potentially scrollable (#10947) 
 
As of this moment Element::ScrollHeight and Element::ScrollWidth are the only places Node::scroll_area is used. Iif I'm not mistaken neither of these functions are used in test-wpt. The most relevant tests I'm aware of are /css-tests/cssom-view-1_dev/scrollWidthHeight*. You can run them with something like the following.

./mach test-css tests/wpt/css-tests/cssom-view-1_dev/html/scrollWidthHeight*
The outcome of these tests shouldn't change since none of the elements have overflow-(x|y): visible. Perhapse a test should be added for an element with overflow-(x|y): visible?

**Comment by broesamle** [comment]
when adding the new test and running `./mach test-css --manifest-update SKIP_TESTS` I get a huge difference in 'tests/wpt/metadata-css/MANIFEST.json.'
(despite it is based on a recent version https://github.com/servo/servo/commit/4d13f9f5d5cbca8de0c950a4d2344fc312a9c186)

785199 vs 1208311 lines!!

difference starting with items that certainly do not belong to my new test file:
``` 
   "items": {
      "reftest": {
        "compositing-1_dev/html/background-blend-mode-gradient-image.htm": [
          {
            "path": "compositing-1_dev/html/background-blend-mode-gradient-image.htm",
            "references": [
              [
                "/compositing-1_dev/html/reference/background-blend-mode-gradient-image-ref.htm",
                "=="
              ]
            ],
            "url": "/compositing-1_dev/html/background-blend-mode-gradient-image.htm"
          }
        ],
        "compositing-1_dev/html/mix-blend-mode-animation.htm": [
          {
            "path": "compositing-1_dev/html/mix-blend-mode-animation.htm",
            "references": [
              [
                "/compositing-1_dev/html/reference/mix-blend-mode-animation-ref.htm",
. . .    
. . .            
```

**Comment by jdm** [comment]
That's a known issue. However, we currently cannot add new tests to the `css-tests` directory; we only mirror tests from upstream so it would be overwritten. It should be added to `tests/wpt/web-platform-tests/cssom-view` instead.

**Comment by broesamle** [comment]
maybe I am missing something but this tiny test should pass?!
```
<!doctype html>
<html>
<script src="/resources/testharness.js" type="text/javascript"></script>
<script src="/resources/testharnessreport.js" type="text/javascript"></script>
<body id="thebody">
</body>
<script>
test(function() {
    document.body.style.overflow = "hidden";
    assert_equals(document.body.style.overflow, "hidden", "Could not set document.body.style.overflow to `hidden`.");
}, "Ensure that style.overflow can be set properly");
</script>
</html>
```
it tells me: `expected **"hidden"** but got **"hidden hidden"**`

**Comment by broesamle** [comment]
Why would it double the **hidden** in the overflow property?

**Comment by jdm** [comment]
Probably because of a bug in our CSS serialization code in cssstyledeclaration.rs .

**Comment by broesamle** [comment]
To me it looks like #9307 is addressing the issue. At least it is related.
It would make sense to fix the style setting issue first, since testing for this one would be difficult without it.
Another small example demonstrating the issue:
```
<!doctype html>
<html>
<body id="thebody">
</body>
<script>
    console.log("A "+document.body.style.overflow);
    document.body.style.overflow = "hidden";
    console.log("B "+document.body.style.overflow);
    console.log("C "+document.body.getAttribute('style'));
    document.body.setAttribute('style','height:30px');
    console.log("D "+document.body.getAttribute('style'));
</script>
</html>
```
Output is:
```
A 
B hidden hidden
C null
D height:30px
```

**Comment by jdm** [comment]
Now that #11377, is the output improved?

**Comment by broesamle** [comment]
thats what i hope too, ill check tomorrow.

But I think it is more related to the test i have in mind. 


Martin Br:osamle     m-broe.de

-------- Ursprüngliche Nachricht --------
Von: Josh Matthews <notifications@github.com> 
Datum: 25.05.2016  17:45  (GMT+01:00) 
An: servo/servo <servo@noreply.github.com> 
Cc: "Martin Br:osamle" <info@m-broe.de>,Comment <comment@noreply.github.com> 
Betreff: Re: [servo/servo] Node::scroll_area should check if the element is potentially scrollable (#10947) 
 
Now that #11377 merged, is the output improved?

**Comment by broesamle** [comment]
With a freshly build version of this morning I still get the style attribute doubled, as before. :-/ 
This attribute issue looks like a separate thing. I did check for other 'suspisious' issues in this respect but did not find any. Shall I open one?

Fixing #10947 in itself should only need half a line of code change but its hard to come up with a test for it as long as setting/getting the style attrib does not work properly.

**Comment by jdm** [comment]
Yes, please do file a separate issue for the attribute one.

**Comment by broesamle** [comment]
here it is #11448

**Comment by broesamle** [comment]
Having learned what shorthands are I could now avoid them in the testing. 
Does it make sense to create a PR and review the test already?
https://github.com/broesamle/servo/commit/364f8084335805d3b1fd846e187cfe153f064f3c

**Comment by broesamle** [comment]
currently I am struggeling with my natural english of not and or:

how is the following phrase in [the spec](https://drafts.csswg.org/cssom-view/#potentially-scrollable) are meant?
> The element’s used value of the `overflow-x` or `overflow-y` properties is not visible.

a) both of them have to be `visible` in order to make this true

b) it is enough if one is `visible` 
(in the sense of: EITHER `overflow-x` is not `visible` OR overflow-Y is not `visible`(OR both are not))


a) `overflow-x==visible`

**Comment by jdm** [comment]
The sentence is considered true if:
* the used value for `overflow-x` is not 'visible' OR
* the used value for `overflow-y` is not 'visible'

**Comment by broesamle** [comment]
during adding the additional condition I find that `potentially_scrollable` always returns true on the example HTML file, which in Firefox provokes the two different behaviours defined in the specs for `scrollHeight`.

Since I now scrambled the logic gatters of my brain back and forth quite a few times I suggest making a PR for the test, review it and then continue on fixing the bug in the servo code. This would ensure that I do not co-evolve a double-negate test and error logic.

**Comment by jdm** [comment]
That sounds like a sensible choice :)

**Comment by broesamle** [comment]
just removed the typos.

**Comment by broesamle** [comment]
It looks like `overflow_x_is_visible` and/or `overflow_y_is_visible` are not carrying the right values.

Any hints for logging the overflow properties are welcome. 

1) I tried `./target/debug/servo -Z dump-flow-tree ...`

should I find something like `overflow=Visible` instead pf loads of `overflow=Overflow` here; or somewhere else?

2) When running `./target/debug/servo -Z dump-display-list ...` on the same file I get a 404 error in the servo window!

**Comment by izgzhen** [comment]
@broesamle Do you mean `element::overflow_x_is_visible`? I experimented with it a bit and looks like it is truly problematic. Also, `doc.is_fully_active` in `element::scroll` is not working as I intended neither.
